File: jdroid-android-about/src/main/java/com/jdroid/android/about/AboutFragment.java
package com.jdroid.android.about;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.ImageView;
import android.widget.TextView;

import com.jdroid.android.about.feedback.RateAppStats;
import com.jdroid.android.activity.ActivityLauncher;
import com.jdroid.android.application.AbstractApplication;
import com.jdroid.android.recycler.AbstractRecyclerFragment;
import com.jdroid.android.recycler.FooterRecyclerViewType;
import com.jdroid.android.recycler.RecyclerViewAdapter;
import com.jdroid.android.recycler.RecyclerViewContainer;
import com.jdroid.android.recycler.RecyclerViewType;
import com.jdroid.android.utils.AppUtils;
import com.jdroid.android.utils.ExternalAppsUtils;
import com.jdroid.java.collections.Lists;
import com.jdroid.java.remoteconfig.RemoteConfigParameter;

import java.util.List;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class AboutFragment extends AbstractRecyclerFragment {

	private List<Object> aboutItems = Lists.newArrayList();

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		// Header
		aboutItems.add("");

		final String website = getWebsite();
		if (website != null) {
			aboutItems.add(new AboutItem(R.drawable.jdroid_ic_website_black_24dp, R.string.jdroid_website) {
				@Override
				public void onSelected(Activity activity) {
					ExternalAppsUtils.openUrl(website);
				}
			});
		}

		final String contactUsEmailAddress = getContactUsEmail();
		if (contactUsEmailAddress != null) {
			aboutItems.add(new AboutItem(R.drawable.jdroid_ic_contact_us_black_24dp, R.string.jdroid_contactUs) {

				@Override
				public void onSelected(Activity activity) {
					if (ExternalAppsUtils.openEmail(contactUsEmailAddress)) {
						AboutAppModule.Companion.get().getModuleAnalyticsSender().trackContactUs();
					} else {
						// TODO Improve this adding a toast or something
					}
				}
			});
		}

		if (AboutAppModule.Companion.get().getAboutContext().getSpreadTheLoveFragmentClass() != null) {
			aboutItems.add(new AboutItem(R.drawable.jdroid_ic_spread_the_love_black_24dp, R.string.jdroid_spreadTheLove) {

				@Override
				public void onSelected(Activity activity) {
					ActivityLauncher.startActivity(activity, SpreadTheLoveActivity.class);
				}
			});
		}
		aboutItems.add(new AboutItem(R.drawable.jdroid_ic_libraries_black_24dp, R.string.jdroid_libraries) {

			@Override
			public void onSelected(Activity activity) {
				ActivityLauncher.startActivity(activity, LibrariesActivity.class);
			}
		});

		RemoteConfigParameter privacyPolicyUrlRemoteConfigParameter = AbstractApplication.get().getAppContext().getPrivacyPolicyUrl();
		if (privacyPolicyUrlRemoteConfigParameter != null) {
			String privacyPolicyUrl = AbstractApplication.get().getRemoteConfigLoader().getString(privacyPolicyUrlRemoteConfigParameter);
			aboutItems.add(new AboutItem(R.drawable.jdroid_ic_privacy_policy_black_24dp, R.string.jdroid_privacyPolicy) {

				@Override
				public void onSelected(Activity activity) {
					ExternalAppsUtils.openUrl(privacyPolicyUrl);
				}
			});
		}

		if (AboutAppModule.Companion.get().getAboutContext().isBetaTestingEnabled()) {
			aboutItems.add(new AboutItem(R.drawable.jdroid_ic_beta_black_24dp, R.string.jdroid_beta) {

				@Override
				public void onSelected(Activity activity) {
					ExternalAppsUtils.openUrl(AboutAppModule.Companion.get().getAboutContext().getBetaTestingUrl());
				}
			});
		}
		aboutItems.addAll(getCustomAboutItems());
	}

	@Override
	public void onViewCreated(View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		RecyclerViewAdapter recyclerViewAdapter = new RecyclerViewAdapter(Lists.newArrayList(new HeaderRecyclerViewType(), new AboutRecyclerViewType()), aboutItems);
		if (rateAppViewEnabled()) {
			recyclerViewAdapter.setFooter(new AboutFooterRecyclerViewType());
		}
		setAdapter(recyclerViewAdapter);
	}

	protected String getWebsite() {
		return AbstractApplication.get().getAppContext().getWebsite();
	}

	protected String getContactUsEmail() {
		return AbstractApplication.get().getAppContext().getContactUsEmail();
	}

	protected List<AboutItem> getCustomAboutItems() {
		return Lists.newArrayList();
	}

	@MainThread
	protected Boolean rateAppViewEnabled() {
		return RateAppStats.displayRateAppView();
	}

	@Override
	protected boolean isDividerItemDecorationEnabled() {
		return true;
	}

	public class HeaderRecyclerViewType extends RecyclerViewType<String, HeaderItemHolder> {

		@Override
		protected Class<String> getItemClass() {
			return String.class;
		}

		@Override
		protected Integer getLayoutResourceId() {
			return R.layout.jdroid_about_header_view;
		}

		@Override
		public RecyclerView.ViewHolder createViewHolderFromView(View view) {
			HeaderItemHolder holder = new HeaderItemHolder(view);
			holder.appIcon = findView(view, R.id.appIcon);
			holder.appName = findView(view, R.id.appName);
			holder.version = findView(view, R.id.version);
			return holder;
		}

		@Override
		public void fillHolderFromItem(String item, HeaderItemHolder holder) {
			holder.appIcon.setImageResource(AbstractApplication.get().getLauncherIconResId());
			holder.appName.setText(AbstractApplication.get().getAppName());
			holder.version.setText(getString(R.string.jdroid_version, AppUtils.getVersionName()));
			if (AbstractApplication.get().getAppContext().isDebugSettingsEnabled()) {
				holder.appIcon.setOnClickListener(new OnClickListener() {

					@Override
					public void onClick(View v) {
						AbstractApplication.get().getDebugContext().launchActivityDebugSettingsActivity(getActivity());
					}
				});
			}
		}

		@NonNull
		@Override
		public RecyclerViewContainer getRecyclerViewContainer() {
			return AboutFragment.this;
		}
	}

	public static class HeaderItemHolder extends RecyclerView.ViewHolder {

		protected ImageView appIcon;
		protected TextView appName;
		protected TextView version;

		public HeaderItemHolder(View itemView) {
			super(itemView);
		}
	}

	public class AboutFooterRecyclerViewType extends FooterRecyclerViewType {

		@Override
		protected Integer getLayoutResourceId() {
			return R.layout.jdroid_about_footer_view;
		}

		@NonNull
		@Override
		public RecyclerViewContainer getRecyclerViewContainer() {
			return AboutFragment.this;
		}
	}

	public class AboutRecyclerViewType extends RecyclerViewType<AboutItem, AboutItemHolder> {

		@Override
		protected Class<AboutItem> getItemClass() {
			return AboutItem.class;
		}

		@Override
		protected Integer getLayoutResourceId() {
			return R.layout.jdroid_default_item;
		}

		@Override
		public RecyclerView.ViewHolder createViewHolderFromView(View view) {
			AboutItemHolder holder = new AboutItemHolder(view);
			holder.image = findView(view, R.id.image);
			holder.name = findView(view, R.id.name);
			return holder;
		}

		@Override
		public void fillHolderFromItem(AboutItem item, AboutItemHolder holder) {
			holder.image.setImageResource(item.getIconResId());
			holder.name.setText(item.getNameResId());
		}

		@Override
		public void onItemSelected(AboutItem item, View view) {
			item.onSelected(getActivity());
		}

		@NonNull
		@Override
		public RecyclerViewContainer getRecyclerViewContainer() {
			return AboutFragment.this;
		}
	}

	public static class AboutItemHolder extends RecyclerView.ViewHolder {

		protected ImageView image;
		protected TextView name;

		public AboutItemHolder(View itemView) {
			super(itemView);
		}
	}
}

File: jdroid-android-about/src/main/java/com/jdroid/android/about/AboutFragment.kt
package com.jdroid.android.about

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.MainThread
import androidx.recyclerview.widget.RecyclerView
import com.jdroid.android.about.feedback.RateAppStats
import com.jdroid.android.activity.ActivityLauncher
import com.jdroid.android.application.AbstractApplication
import com.jdroid.android.recycler.AbstractRecyclerFragment
import com.jdroid.android.recycler.FooterRecyclerViewType
import com.jdroid.android.recycler.RecyclerViewAdapter
import com.jdroid.android.recycler.RecyclerViewContainer
import com.jdroid.android.recycler.RecyclerViewType
import com.jdroid.android.utils.AppUtils
import com.jdroid.android.utils.ExternalAppsUtils

open class AboutFragment : AbstractRecyclerFragment() {

    private val aboutItems = mutableListOf<Any>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Header
        aboutItems.add("")

        val website = getWebsite()
        if (website != null) {
            aboutItems.add(object : AboutItem(R.drawable.jdroid_ic_website_black_24dp, R.string.jdroid_website) {
                override fun onSelected(activity: Activity) {
                    ExternalAppsUtils.openUrl(website)
                }
            })
        }

        val contactUsEmailAddress = getContactUsEmail()
        if (contactUsEmailAddress != null) {
            aboutItems.add(object : AboutItem(R.drawable.jdroid_ic_contact_us_black_24dp, R.string.jdroid_contactUs) {

                override fun onSelected(activity: Activity) {
                    if (ExternalAppsUtils.openEmail(contactUsEmailAddress)) {
                        AboutAppModule.get().getModuleAnalyticsSender().trackContactUs()
                    } else {
                        // TODO Improve this adding a toast or something
                    }
                }
            })
        }

        if (AboutAppModule.get().aboutContext.getSpreadTheLoveFragmentClass() != null) {
            aboutItems.add(object : AboutItem(R.drawable.jdroid_ic_spread_the_love_black_24dp, R.string.jdroid_spreadTheLove) {

                override fun onSelected(activity: Activity) {
                    ActivityLauncher.startActivity(activity, SpreadTheLoveActivity::class.java)
                }
            })
        }
        aboutItems.add(object : AboutItem(R.drawable.jdroid_ic_libraries_black_24dp, R.string.jdroid_libraries) {

            override fun onSelected(activity: Activity) {
                ActivityLauncher.startActivity(activity, LibrariesActivity::class.java)
            }
        })

        val privacyPolicyUrlRemoteConfigParameter = AbstractApplication.get().appContext.privacyPolicyUrl
        if (privacyPolicyUrlRemoteConfigParameter != null) {
            val privacyPolicyUrl = AbstractApplication.get().remoteConfigLoader!!.getString(privacyPolicyUrlRemoteConfigParameter)!!
            aboutItems.add(object : AboutItem(R.drawable.jdroid_ic_privacy_policy_black_24dp, R.string.jdroid_privacyPolicy) {

                override fun onSelected(activity: Activity) {
                    ExternalAppsUtils.openUrl(privacyPolicyUrl)
                }
            })
        }

        if (AboutAppModule.get().aboutContext.isBetaTestingEnabled()) {
            aboutItems.add(object : AboutItem(R.drawable.jdroid_ic_beta_black_24dp, R.string.jdroid_beta) {

                override fun onSelected(activity: Activity) {
                    ExternalAppsUtils.openUrl(AboutAppModule.get().aboutContext.getBetaTestingUrl())
                }
            })
        }
        aboutItems.addAll(getCustomAboutItems())
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val recyclerViewAdapter = RecyclerViewAdapter(listOf(HeaderRecyclerViewType(), AboutRecyclerViewType()), aboutItems)
        if (rateAppViewEnabled()) {
            recyclerViewAdapter.setFooter(AboutFooterRecyclerViewType())
        }
        setAdapter(recyclerViewAdapter)
    }

    protected open fun getWebsite(): String? {
        return AbstractApplication.get().appContext.website
    }

    protected open fun getContactUsEmail(): String? {
        return AbstractApplication.get().appContext.contactUsEmail
    }

    protected open fun getCustomAboutItems(): List<AboutItem> {
        return listOf()
    }

    @MainThread
    protected fun rateAppViewEnabled(): Boolean {
        return RateAppStats.displayRateAppView()
    }

    override fun isDividerItemDecorationEnabled(): Boolean {
        return true
    }

    inner class HeaderRecyclerViewType : RecyclerViewType<String, HeaderItemHolder>() {

        override fun getItemClass(): Class<String> {
            return String::class.java
        }

        override fun getLayoutResourceId(): Int {
            return R.layout.jdroid_about_header_view
        }

        override fun createViewHolderFromView(view: View): RecyclerView.ViewHolder {
            val holder = HeaderItemHolder(view)
            holder.appIcon = findView(view, R.id.appIcon)
            holder.appName = findView(view, R.id.appName)
            holder.version = findView(view, R.id.version)
            return holder
        }

        override fun fillHolderFromItem(item: String, holder: HeaderItemHolder) {
            holder.appIcon.setImageResource(AbstractApplication.get().launcherIconResId)
            holder.appName.text = AbstractApplication.get().appName
            holder.version.text = getString(R.string.jdroid_version, AppUtils.getVersionName())
            if (AbstractApplication.get().appContext.isDebugSettingsEnabled) {
                holder.appIcon.setOnClickListener { AbstractApplication.get().debugContext.launchActivityDebugSettingsActivity(activity) }
            }
        }

        override fun getRecyclerViewContainer(): RecyclerViewContainer {
            return this@AboutFragment
        }
    }

    class HeaderItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        lateinit var appIcon: ImageView
        lateinit var appName: TextView
        lateinit var version: TextView
    }

    inner class AboutFooterRecyclerViewType : FooterRecyclerViewType() {

        override fun getLayoutResourceId(): Int {
            return R.layout.jdroid_about_footer_view
        }

        override fun getRecyclerViewContainer(): RecyclerViewContainer {
            return this@AboutFragment
        }
    }

    inner class AboutRecyclerViewType : RecyclerViewType<AboutItem, AboutItemHolder>() {

        override fun getItemClass(): Class<AboutItem> {
            return AboutItem::class.java
        }

        override fun getLayoutResourceId(): Int {
            return R.layout.jdroid_default_item
        }

        override fun createViewHolderFromView(view: View): RecyclerView.ViewHolder {
            val holder = AboutItemHolder(view)
            holder.image = findView(view, R.id.image)
            holder.name = findView(view, R.id.name)
            return holder
        }

        override fun fillHolderFromItem(item: AboutItem, holder: AboutItemHolder) {
            holder.image.setImageResource(item.iconResId)
            holder.name.setText(item.nameResId)
        }

        public override fun onItemSelected(item: AboutItem, view: View) {
            item.onSelected(activity)
        }

        override fun getRecyclerViewContainer(): RecyclerViewContainer {
            return this@AboutFragment
        }
    }

    class AboutItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        lateinit var image: ImageView
        lateinit var name: TextView
    }
}
